### ISSUE TITLE
Unnecessary double-quoting of MIME header parameters

### ISSUE THREAD

**Issue by wil** [mailgun/flanker#77]
Given a MIME header as follows:

```Content-Type: multipart/alternative; boundary=decafbad```

Doing a round-trip parse-then-serialize using Flanker will result in the header being represented as:

```Content-Type: multipart/alternative; boundary="decafbad"```

This is due to `flanker.mime.messages.headers.encoding.encode_param` calling `email.message._formatparam` without explicitly specifying the optional `quote` argument (which defaults to `True`.)

```python
>>> email.message._formatparam('boundary', 'hello')
'boundary="hello"'
>>> email.message._formatparam('boundary', 'hello', quote=False)
'boundary=hello'

>>> email.message._formatparam('boundary', 'hel<lo')
'boundary="hel<lo"'
>>> email.message._formatparam('boundary', 'hel<lo', quote=False)
'boundary="hel<lo"'
```

As seen above, if Flanker passes `quote=False` when calling `email.message._formatparam`, the parameter value would only be quoted if the value does not contain any special characters.

AFAICT, header parameters can be always double-quoted (even if the value is a legal "token", and contains no special characters -- as the example above shows.) So, this is arguably not a bug.

However, I have found (empirically) that Gmail tends to generate hex boundary values, and do not double-quote it in the `Content-Type` header. This causes DKIM signature (which covers the `Content-Type` header) verification to fail on messages that have gone through this roundtrip conversion.

We have found this to be an issue with using Mailgun API to retrieve inbound messages with the `Accept: message/rfc2822` header. We found that the message returned has an incorrectly re-encoded `Content-Type` header `boundary` parameter. This in turn causes DKIM verification to fail. I'm mentioning this because we think that Mailgun is using Flanker internally.

Obviously, the "fix" above could break in the opposite scenario where the original header had a double-quoted `boundary` parameter value even when it contains no special characters. Unfortunately, I do not have evidence whether such messages are out there in the wild, and if so, how frequent.

I also realise that there are other ways that DKIM verification could fail when doing such roundtrip conversions. However, at least for a big major email provider (Gmail), fixing this header value encoding takes care of the issue in our testing.

**Comment by jeremyschlatter** [comment]
Thanks for pointing this out!

> Obviously, the "fix" above could break in the opposite scenario where the original header had a double-quoted boundary parameter value even when it contains no special characters. Unfortunately, I do not have evidence whether such messages are out there in the wild, and if so, how frequent.

I checked this for a few thousand messages in the wild, and this scenario (double-quoted parameter with no special characters) actually represented the majority of cases -- 3901 out of 4844. Turns out this is especially common for charset declarations, like `Content-Type: text/plain; charset="utf-8"`. Given this finding, I don't want to set `quote=False` by default.

The next-best fix would be to track whether or not the parameter was quoted when we parse it, and then use that information when we serialize again. But I don't think that's a trivial change, and there are other places where parse-then-serialize changes the text. I think Flanker probably isn't the place to fix the DKIM verification fail.
